osh/hub/scripts/osh-worker-manager

```diff
@@ -0,0 +1,208 @@
+"""
+
+import os
+import socket
+import subprocess
+import sys
+import tempfile
+
+# TODO: We may use python3 logger instead of `print()` statements for debugging.
+
+
+# On AWS, public hostname is accessible through this command on the vm:
+# `curl http://169.254.169.254/latest/meta-data/public-hostname`
+# TODO: Shall we use that instead of using ip address?
+def get_hostname_from_ip(ip_address):
+    print(f"Getting hostname for {ip_address}")
+    hostname = socket.gethostbyaddr(ip_address)[0]
+    return hostname
+
+
+def create_worker_in_kobo(worker_name):
+    from django.core.exceptions import ObjectDoesNotExist
+    from kobo.hub import models
+
+    # Check if a worker with this name already exists
+    try:
+        kobo_worker = models.Worker.objects.get(name=worker_name)
+    except ObjectDoesNotExist:
+        # If not, create a new one
+        kobo_worker = models.Worker.create_worker(worker_name)
+
+    # TODO: Detect worker architecture automatically.
+
+        kobo_worker.arches.add(models.Arch.objects.get(name="x86_64"))
+        kobo_worker.arches.add(models.Arch.objects.get(name="noarch"))
+    except ObjectDoesNotExist:
+        print("x86_64 or noarch do not exist in Arch model.", file=sys.stderr)
+        kobo_worker.delete()
+        sys.exit(1)
+
+    try:
+        kobo_worker.channels.add(models.Channel.objects.get(name="default"))
+    except ObjectDoesNotExist:
+        print("Default channel does not exist.", file=sys.stderr)
+        kobo_worker.delete()
+        sys.exit(1)
+
+    # Each single use virtual machine allocated should get exactly one task
+    # and it should be deleted after that.
+    kobo_worker.max_load = 1
+
+    kobo_worker.save()
+    print(f"Created kobo worker for hostname {worker_name}")
+    return kobo_worker
+
+
+def start_worker(kobo_worker):
+    """
+    This function works in below steps:
+    1. Generate a `worker.conf` with the newly generated `worker_key`.
+    2. Copy the `worker.conf` to the worker.
+    3. Move the file to `/etc/osh/worker.conf` and change it's permissions to `root:root`.
+    4. Start the `osh-worker` service on the virtual machine.
+    """
+    # This function looks like a hack and we should investigate if it could be done better.
+    # For example, through ansible or paramiko.
+
+    from django.conf import settings
+    from django.template.loader import render_to_string
+
+    # Generate /etc/osh/worker.conf through jinja template
+    worker_conf = render_to_string("worker.conf.j2",
+                                   {"SINGLE_USE_WORKER_OSH_HUB_URL": settings.SINGLE_USE_WORKER_OSH_HUB_URL,
+                                    "OSH_WORKER_KEY": kobo_worker.worker_key})
+
+    # Create a temporary file and save the template
+    with tempfile.NamedTemporaryFile() as temporary_file:
+        temporary_file.write(worker_conf.strip().encode())
+        temporary_file.flush()
+        # File is closed, but not removed.
```

Comment:
This does not work as documented in my testing and I get below error if I close the file here:

```
stat local "/tmp/tmprimblhg2": No such file or directory^M
Failed to copy worker.conf to new worker ec2-54-184-164-118.us-west-2.compute.amazonaws.com
```

Reply:
`temporary_file.flush()` does **not** close the file.  I think it will be closed at the and of the `with ... as` block.  The fact whether the file will be deleted on close is controlled by the `delete_on_close ` argument of `tempfile.NamedTemporaryFile()` and it defaults to `True`.  So I would say it works as documented: https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile

Reply:
@siteshwar What exactly does not work as documented?

Reply:
> @siteshwar What exactly does not work as documented?

I was not setting `delete_on_close=False`. So I missed that in the documentation.

Reply:
On another note, `delete_on_close` is [python 3.12](https://docs.python.org/3.12/whatsnew/3.12.html#tempfile). Earlier versions only have `delete`.